Add `#[pyclass(generic)]`

IvanIsCoding opened the pull request:
Closes #4849

Follow up of #4917

* [x] Add more tests 
* [x] Add changelog entry
* [x]  Update the documentation
* [x] Reject `generic` for enums; let's focus only on supporting structs for now
* [x] ~Error in Python 3.7 and 3.8~ (it just fails to compile for now)
* [x] Fix the compile error tests regressions; I haven't investigated why they fail yet

edit: it is mostly ready to review now

IvanIsCoding commented on pyo3-macros-backend/src/pyclass.rs:
```diff
@@ -2023,6 +2030,45 @@ fn pyclass_hash(
     }
 }
 
+fn pyclass_class_geitem(
```
@davidhewitt do you thinnk this is a step towards the right direction? I was in doubt where is the easiest place to inject a `@classmethod` in the macros

IvanIsCoding replied:
[mejrs](https://github.com/mejrs) feedback fixed my issue

mejrs commented:
Assuming we want to go with `#[pyclass(generic)]` as the syntax, I think this is roughly the best way to insert that method.

As for the error: what this builder does is that it a) emits an `extern "C" fn` that wraps a rust function, and b) emits code that gives Python a pointer to the function and its name. It does not emit the rust function; you have to add that somewhere. Probably something like this:

https://github.com/PyO3/pyo3/blob/f2a8460559d8f58269f306679fc58c9d101f70cf/pyo3-macros-backend/src/pyclass.rs#L456-L462

This should also prevent the user from both writing `#[pyclass(generic)]` and creating a method named `__class_getitem__` (please add a test for this)

IvanIsCoding commented:
I added tests and updated the documentation. I am confident it all works. For compilation failure, I tested users redefining `__class_geitem__` and blocked enums for the time being.

I am still not sure how to block `#[pyclass(generic)]` for Python 3.7 and Python 3.8. I don't know if we should do it, but currently specifying it will cause compilation failures anyway as #4917 is gated for Python 3.9+ anyway

mejrs commented on tests/ui/invalid_pyclass_generic.rs:
```diff
@@ -0,0 +1,24 @@
+    }
+}
+
+fn main() {}
```
Please add an empty line beyond this

IvanIsCoding replied:
Done.

mejrs commented on tests/ui/pyclass_generic_enum.rs:
```diff
@@ -0,0 +1,15 @@
+    B { y: f64, z: f64 },
+}
+
+fn main() {}
```
Same thing; please add a trailing newline.

IvanIsCoding replied:
Done.

mejrs reviewed (approved):
Sorry for the long wait, this looks good to me. Please add some newlines though :)

IvanIsCoding commented:
I need to rerun the tests with `TRYBUILD=overwrite` to generate the latest error messages, I think the CI failure should be easy to fix.

And no worries about the delay, this is like syntax-sugar. I am happy the earlier PR got merged fast and released in 0.24, my project is already using it (https://github.com/Qiskit/rustworkx/pull/1421).

mejrs commented on guide/pyclass-parameters.md:
```diff
@@ -23,6 +23,7 @@
 | `subclass` | Allows other Python classes and `#[pyclass]` to inherit from this class. Enums cannot be subclassed. |
 | `unsendable` | Required if your struct is not [`Send`][params-3]. Rather than using `unsendable`, consider implementing your struct in a thread-safe way by e.g. substituting [`Rc`][params-4] with [`Arc`][params-5]. By using `unsendable`, your class will panic when accessed by another thread. Also note the Python's GC is multi-threaded and while unsendable classes will not be traversed on foreig ...
 | `weakref` | Allows this class to be [weakly referenceable][params-6]. |
+| `generic` | Implements runtime parametrization for the class following [PEP 560](https://peps.python.org/pep-0560/). |
```
Just one more thing 😅 . This table is supposed to be in alphabetical order.

IvanIsCoding replied:
Done.

IvanIsCoding commented:
Ok CI is green now, I moved the table entry to be in alphabetical order as well.

IvanIsCoding commented:
I will try to fix the Python 3.9 failure tomorrow

IvanIsCoding commented:
@mejrs https://docs.python.org/3/library/typing.html#typing.TypeAlias according to the Python docs `typing.TypeAlias` is only available in Python 3.10+. So I updated the tests to run only on 3.10 onwards.

Ideally we would use the `type` statement but that is only on 3.12+. So it is what it is.

davidhewitt reviewed (approved):
Thanks!
